server/app.js

```diff
@@ -0,0 +1,250 @@
+      // considered as the controller for the given path. This options argument is
+      // therefore removed when sending it to the newInstanceMethods[method]
+      // because the express application expects a function
+      const [secondLastArg] = middlewareArgs.slice(-2);
+      middlewareArgs = middlewareArgs.slice(0, middlewareArgs.length - 1);
+      handleLastArg(secondLastArg, firstArg, method, lastArg);
+    }
+  }
+  // This needs to return the output of the bound method's function
+  // Do not return before this stage
+  return newInstanceMethods[method](...middlewareArgs);
+};
+
+/**
+ * Replaces the methods of the given instance with that which will allow us to capture the route
+ * definition and document it with swagger
+ * The app methods being replaced include app.all, app.use, app.[HTTP method] e.g, app.get
+ * @param {express} appInstance The express application instance whose methods need to be replaced
+ * @returns {Promise<void>} A promise that resolves after the methods
+ * have been successfully replaced */
+const replaceMethods = async (appInstance) => {
+  // The new instance methods that point to the original function definitions for the app instance
+  // Here we bind the appInstance, so that "this" refers to the
+  // appInstance when the method is called
+  const newInstanceMethods = {
+    use: appInstance.use.bind(appInstance),
+    all: appInstance.all.bind(appInstance),
+  };
+
+  // Loop one by one adding the original function definitions to the newInstanceMethods object
+  // only if the appInstance exposes the said HTTP method
+  await loopSequentially(httpMethods, (method) => {
+    newInstanceMethods[method] = appInstance[method] && appInstance[method].bind(appInstance);
+  });
+
+  // Loop one by one replacing the appInstance's methods found in the keys of the
+  // newInstanceMethods with our own function that passes the arguments to our swagger
+  // handler which will then handle the addition of the paths, if any, to the swagger schema
+  await loopSequentially(Object.keys(newInstanceMethods), (method) => {
+    if (method) {
+      /* eslint-disable-next-line no-param-reassign */
+      appInstance[method] = (...args) => swaggerHandler(newInstanceMethods, method, ...args);
+    }
+  });
+};
+
+/**
+ * This method set's up all middleware and routes needed by the express app instance
+ * @return {Promise<void>} Resolves when the set up is complete
+ */
+export const setUp = async () => {
+  app.schema = swaggerConfig;
+
+  await replaceMethods(app);
+
+  // Log requests to the console.
+  app.use(logger('dev'));
+
+  app.use(bodyParser.json());
+  app.use(bodyParser.urlencoded({ extended: false }));
+
+  // Register the routes
+  routes(app);
+
+  // Here set the swagger docs then handle 404s
+  // It shows an example of how you can add swagger doc specification using an options object.
+  app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(app.schema), {
```

Comment:
@davenjeru why is this being done both in the `app.js` file and `index.js` file?

Reply:
So, The app schema changes in the callback for app.listen in app.js when defining the servers, so it needs the new app.schema to serve it.
It is also defined in index.js just before the `app.use(*)` because if we don't, `/api-docs` will be handled by the handler for `app.use(*)` and we don't want that.